db/db_secondary_test.cc

```diff
@@ -0,0 +1,217 @@
+       {"VersionSet::ProcessManifestWrites:AfterNewManifest",
+        "VersionSet::MaybeSwitchManifest:AfterGetCurrentManifestPath:1"}});
+  SyncPoint::GetInstance()->EnableProcessing();
+
+  // Make sure db calls RecoverLogFiles so as to trigger a manifest write,
+  // which causes the db to switch to a new MANIFEST upon start.
+  port::Thread ro_db_thread([&]() {
+    DB* db_secondary = nullptr;
+    Options options1;
+    options1.env = env_;
+    options1.max_open_files = -1;
+    Status s = DB::OpenAsSecondary(options1, dbname_, secondary_dbname_,
+                                   &db_secondary);
+    ASSERT_OK(s);
+    delete db_secondary;
+  });
+  Reopen(options);
+  ro_db_thread.join();
+  Close();
+}
+
+TEST_F(DBSecondaryTest, MissingTableFileDuringOpen) {
+  Options options;
+  options.env = env_;
+  options.level0_file_num_compaction_trigger = 4;
+  Reopen(options);
+  for (int i = 0; i != options.level0_file_num_compaction_trigger; ++i) {
+    ASSERT_OK(Put("foo", "foo_value" + std::to_string(i)));
+    ASSERT_OK(Put("bar", "bar_value" + std::to_string(i)));
+    ASSERT_OK(dbfull()->Flush(FlushOptions()));
+  }
+  ASSERT_OK(dbfull()->TEST_WaitForFlushMemTable());
+  ASSERT_OK(dbfull()->TEST_WaitForCompact());
+  DB* db1 = nullptr;
+  Options options1;
+  options1.env = env_;
+  options1.max_open_files = -1;
+  Status s = DB::OpenAsSecondary(options1, dbname_, secondary_dbname_, &db1);
+  ASSERT_OK(s);
+  ReadOptions ropts;
+  ropts.verify_checksums = true;
+  std::string value;
+  ASSERT_OK(db1->Get(ropts, "foo", &value));
+  ASSERT_EQ("foo_value" +
+                std::to_string(options.level0_file_num_compaction_trigger - 1),
+            value);
+  ASSERT_OK(db1->Get(ropts, "bar", &value));
+  ASSERT_EQ("bar_value" +
+                std::to_string(options.level0_file_num_compaction_trigger - 1),
+            value);
+  delete db1;
+  Close();
+}
+
+TEST_F(DBSecondaryTest, MissingTableFile) {
+  Options options;
+  options.env = env_;
+  options.level0_file_num_compaction_trigger = 4;
+  Reopen(options);
+
+  DB* db1 = nullptr;
+  Options options1;
+  options1.env = env_;
+  options1.max_open_files = -1;
+  Status s = DB::OpenAsSecondary(options1, dbname_, secondary_dbname_, &db1);
+  ASSERT_OK(s);
+
+  for (int i = 0; i != options.level0_file_num_compaction_trigger; ++i) {
+    ASSERT_OK(Put("foo", "foo_value" + std::to_string(i)));
+    ASSERT_OK(Put("bar", "bar_value" + std::to_string(i)));
+    ASSERT_OK(dbfull()->Flush(FlushOptions()));
+  }
+  ASSERT_OK(dbfull()->TEST_WaitForFlushMemTable());
+  ASSERT_OK(dbfull()->TEST_WaitForCompact());
+
+  auto db_secondary = static_cast<DBImplSecondary*>(db1);
+  ASSERT_NE(nullptr, db_secondary);
+  ReadOptions ropts;
+  ropts.verify_checksums = true;
+  std::string value;
+  ASSERT_NOK(db_secondary->Get(ropts, "foo", &value));
+  ASSERT_NOK(db_secondary->Get(ropts, "bar", &value));
+
+  ASSERT_OK(db_secondary->TryCatchUpWithPrimary());
```

Comment:
I'm confused by this interface. TryCatchUpWithPrimary() is not a public function, so who is supposed to call it when users run a secondary?

Reply:
I'm going to add a thread calling this function periodically to update version storage info. It's not public intentionally.

Reply:
Update: this function is now public.